new-log-viewer/src/services/decoders/JsonlDecoder.ts

```diff
@@ -0,0 +1,240 @@
+        if ("number" !== typeof timestamp && "string" !== typeof timestamp) {
+            timestamp = INVALID_TIMESTAMP_VALUE;
+        }
+        const dayjsObj: dayjs.Dayjs = dayjs.utc(timestamp);
+        const formattedDate = dayjsObj.format(this.#dateFormat);
+        input = input.replace(this.#datePattern, formattedDate);
+
+        return [
+            dayjsObj.valueOf(),
+            input,
+        ];
+    }
+
+    /**
+     * Replaces placeholders in the input string with corresponding properties from the logEvent
+     * object.
+     *
+     * @param input The input string with placeholders to be replaced.
+     * @param logEvent The log event object containing properties to replace the placeholders.
+     * @return - The input string with placeholders replaced by corresponding property values.
+     * @private
+     */
+    #formatVariables (input: string, logEvent: JsonObject) {
+        // Replace each placeholder with the corresponding property from logEvent
+        for (const propName of this.#propertyNames) {
+            if (propName in logEvent) {
+                const placeholder = `%${propName}`;
+                const propValue = logEvent[propName];
+                let propValueString: string;
+                if ("undefined" === typeof propValue || null === propValue) {
+                    console.error(`Unexpected undefined logEvent[${propName}]`);
+                    propValueString = "undefined";
+                } else if ("object" === typeof propValue) {
+                    propValueString = JSON.stringify(propValue);
+                } else {
+                    propValueString = propValue.toString();
+                }
+                input = input.replace(placeholder, propValueString);
+            }
+        }
+
+        input += "\n";
+
+        return input;
+    }
+
+    /**
+     * Extracts the verbosity from the given log event.
+     *
+     * @param logEvent The log event containing the verbosity.
+     * @return The extracted verbosity.
+     */
+    #extractVerbosity (logEvent: JsonObject) {
+        let verbosity = LOG_VERBOSITY.NONE;
+        const verbosityString: string = logEvent[this.#verbosityPropName] as string;
+        if (false === (verbosityString in LOG_VERBOSITY)) {
+            console.error(`Unable to find verbosity from key ${this.#verbosityPropName}` +
+                ` of type ${typeof verbosityString}`);
+        } else {
+            verbosity = LOG_VERBOSITY[verbosityString as (keyof typeof LOG_VERBOSITY)];
+        }
+
+        return verbosity;
+    }
+
+    constructor (dataArray: Uint8Array) {
+        this.#dataArray = dataArray;
+    }
+
+    setDecodeOptions (options: JsonlDecodeOptionsType): boolean {
+        this.#setTextPattern(
+            options.textPattern
+        );
+        this.#timestampPropName = options.timestampPropName;
+        this.#verbosityPropName = options.verbosityPropName;
+
+        return true;
+    }
+
+    buildIdx (): number {
```

Comment:
Not necessarily. If we enable the optimization in the future where we allow building partial index wnen the decoder `.decode([], startIdx=1, endIdx=any)`. `buildIdx()` can be called after the first page is decoded.

Reply:
Yes.

With some other thoughts, I also agree we want to hide this interface. For now, We can make `buildIdx()` a helper of `decode()` to be invoked conditionally. Then when we work on the optimization we can extract the core logics from `buildIdx()` to allow those to be reused.

Reply:
Sorry, we might not want to do that since we do need the total number of events to see where's the last page, before we can introduce the optimization where we skip building index for the whole log file when only the first page is requested.

Reply:
> for CLP formats, I would deserialize + decode the preamble in the constructor.

Sure. Let me correct the docs.